File: challenge/leet/easy/q599.py
import unittest
from typing import List


class Solution:
    def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
        dp = dict(map(lambda x: (x[1], x[0]), enumerate(list1)))
        res, low = [], float('inf')
        for x, v in enumerate(list2):
            if v in dp:
                lo = x + dp[v]
                if lo < low:
                    low = lo
                    res = [v]
                elif lo == low:
                    res.append(v)
        return res


class MyTestCase(unittest.TestCase):
    def test_something(self):
        s = Solution()

        self.assertEqual(["Shogun"],
           s.findRestaurant(["Shogun","Tapioca Express","Burger King","KFC"],
                            ["Piatti","The Grill at Torrey Pines","Hungry Hunter Steakhouse","Shogun"]))


if __name__ == '__main__':
    unittest.main()

File: challenge/leet.gcc/easy/q599/q599.cpp
/**
 * 599. Minimum Index Sum of Two Lists
 * https://leetcode.com/problems/minimum-index-sum-of-two-lists/description/
 */

#include <vector>
#include <map>
#include <string>
#include <iostream>
#include <numeric>

using namespace std;

class Solution {
public:
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        map<string, int> dp;

        for (int i = 0; i < list1.size(); i++) {
            dp[list1[i]] = i;
        }

        int mx = INT32_MAX;
        vector<string> ans;
        for (int i = 0; i < list2.size(); i++) {
            if (dp.count(list2[i])) {
                int ix = dp[list2[i]] + i;
				if (ix < mx) {
					mx = ix;
                    ans = vector<string>{};
					ans.push_back(list2[i]);
				}
				else if (ix == mx) {
					ans.push_back(list2[i]);
				}
            }
        }

        return ans;
    }
};

int main() {
    Solution s;
    vector<string> u, v, w;

    u = vector<string>{ "Shogun","Tapioca Express","Burger King","KFC" };
    v = vector<string>{ "Piatti","The Grill at Torrey Pines","Hungry Hunter Steakhouse","Shogun" };
    w = s.findRestaurant(u, v);
    cout << "[Shogun]?" << endl << "[" << reduce(next(w.begin()), w.end(), w[0], [](string a, string b) {return move(a) + ", " + b; }) << "]" << endl;

    u = vector<string>{ "Shogun","Tapioca Express","Burger King","KFC" };
    v = vector<string>{ "KFC","Shogun","Burger King" };
    w = s.findRestaurant(u, v);
    cout << "[Shogun]?" << endl << "[" << reduce(next(w.begin()), w.end(), w[0], [](string a, string b) {return move(a) + ", " + b; }) << "]" << endl;

    u = vector<string>{ "happy","sad","good" };
    v = vector<string>{ "sad","happy","good" };
    w = s.findRestaurant(u, v);
    cout << "[sad, happy]?" << endl << "[" << reduce(next(w.begin()), w.end(), w[0], [](string a, string b) {return move(a) + ", " + b; }) << "]" << endl;
}
